### ISSUE TITLE
Concurrent Lumi mode with noRunLumiSort option, bug

### ISSUE THREAD

**Issue by wddgit** [cms-sw/cmssw#36967]
There is a bug in the code implementing the noRunLumiSort configuration option of PoolSource. I can demonstrate the bug in the following test.

First run the following sequence of existing unit tests:

```
   cmsRun FWCore/Integration/test/testRunMergePROD1_cfg.py
   cmsRun FWCore/Integration/test/testRunMergePROD2_cfg.py
   cmsRun FWCore/Integration/test/testRunMergePROD3_cfg.py
   cmsRun FWCore/Integration/test/testRunMergePROD4_cfg.py
   cmsRun FWCore/Integration/test/testRunMergePROD5_cfg.py
   cmsRun FWCore/Integration/test/testRunMergeMERGE_cfg.py
```

Then read the output using this configuration:

```
import FWCore.ParameterSet.Config as cms
 
process = cms.Process("TEST")
 
process.source = cms.Source("PoolSource",
    fileNames = cms.untracked.vstring(
        'file:testRunMerge.root'
    ),
    noRunLumiSort = cms.untracked.bool(True)
)
```

This last step will fail on an assert in the Framework code.
I am still investigating this. It appears to be related to the handling of
run entries in the function 

```
IndexIntoFile::IndexIntoFileItrEntryOrder::IndexIntoFileItrEntryOrder
```

This code with the bug was merged into CMSSW on 12 Nov 2021.

I noticed this while working on the new concurrent runs feature. So far no user complaints. I am
working on a fix and will submit a PR as soon as possible.

**Comment by makortel** [comment]
assign core

**Comment by wddgit** [comment]
@makortel @Dr15Jones 

With an EntryOrder iteration, what do we want the behavior to be when a cmsRun job ends before processing an entire file? This could be caused by a limit on the number of events or lumis. We could sometimes have events written to the file without having a corresponding run or a corresponding lumi written into the file because the run or lumi was expected to be "continued" later, but we never got to the point in the input where it would be continued and written into the output file.

Here are some possible options:

1. Don't allow configuration options that limit the number of events or lumis when using an entryOrder iteration.

2. Extend the logic that cleans up runs and lumis when the process is ending so that any lumis and runs that were started but expected to be "continued" are found, processed, and written if the run or lumi was started (some extra development, not trivial but possible).

3. Ignore it. I am not sure what happens. I didn't test this yet. I'd guess a missing lumi would just cause the next process to ignore the events in that lumi silently and a missing run would lead to an assert failure.

**Comment by makortel** [comment]
@wddgit I think that makes sense. The noRunLumiSort was intended for Merge jobs, that shouldn't set a limit for number of events or lumis.

**Comment by wddgit** [comment]
Resolved by #37532
